File: src/kernel/arch/gem5/Platform.cc
/*
 * Copyright (C) 2016-2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include "mem/MainMemory.h"
#include "mem/MemoryModule.h"
#include "pes/VPE.h"
#include "Args.h"
#include "TCU.h"
#include "Platform.h"

#include <memory>

namespace kernel {

m3::BootInfo::Mod *Platform::_mods;
m3::PEDesc *Platform::_pes;
m3::BootInfo::Mem *Platform::_mems;
m3::BootInfo Platform::_info;

// note that we currently assume here, that compute PEs and memory PEs are not mixed
static peid_t last_pe_id;

void Platform::init() {
    m3::BootInfo *info = &Platform::_info;
    // read kernel env
    m3::GlobAddr kenv(m3::env()->kenv);
    TCU::read_mem(kenv.pe(), kenv.offset(), info, sizeof(*info));

    // read boot modules
    m3::GlobAddr kenvmods(kenv + sizeof(*info));
    size_t mod_size = info->mod_count * sizeof(m3::BootInfo::Mod);
    Platform::_mods = new m3::BootInfo::Mod[info->mod_count];
    TCU::read_mem(kenvmods.pe(), kenvmods.offset(), Platform::_mods, mod_size);

    // read PE descriptions
    m3::GlobAddr kenvpes(kenvmods + mod_size);
    size_t pe_size = sizeof(m3::PEDesc) * info->pe_count;
    Platform::_pes = new m3::PEDesc[info->pe_count];
    TCU::read_mem(kenvpes.pe(), kenvpes.offset(), Platform::_pes, pe_size);

    // read memory regions
    m3::GlobAddr kenvmems(kenvpes + pe_size);
    size_t mem_size = sizeof(m3::BootInfo::Mem) * info->mem_count;
    Platform::_mems = new m3::BootInfo::Mem[info->mem_count];
    TCU::read_mem(kenvmems.pe(), kenvmems.offset(), Platform::_mems, mem_size);

    // build new info for user PEs
    m3::BootInfo uinfo;
    memcpy(&uinfo, info, sizeof(uinfo));
    auto umems = std::make_unique<m3::BootInfo::Mem[]>(info->mem_count);
    auto upes = std::make_unique<m3::BootInfo::PE[]>(info->pe_count - 1);

    // register memory modules
    size_t umemidx = 0, kmemidx = 0;
    size_t peidx = 0;
    MainMemory &mem = MainMemory::get();
    for(size_t i = 0; i < info->pe_count; ++i) {
        m3::PEDesc pedesc = Platform::_pes[i];
        if(pedesc.type() == m3::PEType::MEM) {
            // the first memory module hosts the FS image and other stuff
            if(umemidx == 0) {
                size_t avail = _mems[kmemidx].size();
                if(avail <= Args::kmem)
                    PANIC("Not enough DRAM for kernel memory (" << Args::kmem << ")");
                size_t used = pedesc.mem_size() - avail;
                mem.add(new MemoryModule(MemoryModule::OCCUPIED, m3::GlobAddr(i, 0), used));
                umems[umemidx++] = m3::BootInfo::Mem(0, used, true);

                mem.add(new MemoryModule(MemoryModule::KERNEL, m3::GlobAddr(i, used), Args::kmem));

                mem.add(new MemoryModule(MemoryModule::USER, m3::GlobAddr(i, used + Args::kmem), avail));
                umems[umemidx++] = m3::BootInfo::Mem(used + Args::kmem, avail - Args::kmem, false);
            }
            else {
                if(umemidx >= info->mem_count)
                    PANIC("Not enough memory slots in boot info");

                mem.add(new MemoryModule(MemoryModule::USER, m3::GlobAddr(i, 0), pedesc.mem_size()));
                umems[umemidx++] = m3::BootInfo::Mem(0, pedesc.mem_size(), false);
            }
            kmemidx++;
        }
        else {
            if(kmemidx > 0)
                PANIC("All memory PEs have to be last");
            last_pe_id = i;

            // don't hand out the kernel PE
            if(i > 0) {
                assert(kernel_pe() == 0);
                upes[peidx].id = i;
                upes[peidx].desc = pedesc;
                peidx++;
            }
        }
    }

    // write-back boot info
    uinfo.pe_count = peidx;
    uinfo.mem_count = umemidx;
    TCU::write_mem(kenv.pe(), kenv.offset(), &uinfo, sizeof(uinfo));
    // write-back user PEs
    TCU::write_mem(kenv.pe(), kenvpes.offset(),
                   upes.get(), sizeof(m3::BootInfo::PE) * uinfo.pe_count);
    // write-back user memory regions
    TCU::write_mem(kenv.pe(), kenvpes.offset() + sizeof(m3::BootInfo::PE) * uinfo.pe_count,
                   umems.get(), sizeof(m3::BootInfo::Mem) * uinfo.mem_count);
}

void Platform::add_modules(int, char **) {
    // unused
}

m3::GlobAddr Platform::info_addr() {
    return m3::GlobAddr(m3::env()->kenv);
}

peid_t Platform::kernel_pe() {
    // gem5 initializes the peid for us
    return m3::env()->pe_id;
}
peid_t Platform::first_pe() {
    return m3::env()->pe_id + 1;
}
peid_t Platform::last_pe() {
    return last_pe_id;
}

bool Platform::is_shared(peid_t no) {
    return pe(no).is_programmable();
}

goff_t Platform::rbuf_pemux(peid_t no) {
    if(pe(no).has_virtmem())
        return PEMUX_RBUF_PHYS;
    else
        return PEMUX_RBUF_SPACE;
}

}

File: src/kernel/src/arch/kachel/platform.rs
/*
 * Copyright (C) 2020-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use base::cell::StaticCell;
use base::cfg;
use base::col::{String, Vec};
use base::envdata;
use base::goff;
use base::kif::{self, boot, Perm, TileDesc, TileISA, TileType};
use base::mem::{size_of, GlobAddr};
use base::tcu::{ActId, EpId, TileId, TCU, UNLIM_CREDITS};
use base::vec;

use crate::args;
use crate::ktcu;
use crate::mem::{self, MemMod, MemType};
use crate::platform::{self, tile_desc};
use crate::tiles::KERNEL_ID;

static LAST_TILE: StaticCell<TileId> = StaticCell::new(0);

pub fn init(_args: &[String]) -> platform::KEnv {
    // read kernel env
    let addr = GlobAddr::new(envdata::get().kenv);
    let mut offset = addr.offset();
    let info: boot::Info = ktcu::read_obj(addr.tile(), offset);
    offset += size_of::<boot::Info>() as goff;

    // read boot modules
    let mut mods: Vec<boot::Mod> = vec![boot::Mod::default(); info.mod_count as usize];
    ktcu::read_slice(addr.tile(), offset, &mut mods);
    offset += info.mod_count as goff * size_of::<boot::Mod>() as goff;

    // read tiles
    let mut tiles: Vec<TileDesc> = vec![TileDesc::default(); info.tile_count as usize];
    ktcu::read_slice(addr.tile(), offset, &mut tiles);
    offset += info.tile_count as goff * size_of::<TileDesc>() as goff;

    // read memory regions
    let mut mems: Vec<boot::Mem> = vec![boot::Mem::default(); info.mem_count as usize];
    ktcu::read_slice(addr.tile(), offset, &mut mems);

    // build new info for user tiles
    let mut uinfo = boot::Info {
        mod_count: info.mod_count,
        tile_count: info.tile_count,
        mem_count: info.mem_count,
        serv_count: 0,
    };

    let mut umems = Vec::new();
    let mut utiles = Vec::new();

    // register memory modules
    let mut kmem_idx = 0;
    let mut mem = mem::borrow_mut();
    for (i, tile) in tiles.iter().enumerate() {
        if tile.tile_type() == TileType::MEM {
            // the first memory module hosts the FS image and other stuff
            if kmem_idx == 0 {
                let avail = mems[kmem_idx].size();
                if avail <= args::get().kmem as goff {
                    panic!("Not enough DRAM for kernel memory ({})", args::get().kmem);
                }

                // file system image
                let mut used = tile.mem_size() as goff - avail;
                mem.add(MemMod::new(MemType::OCCUPIED, i as TileId, 0, used));
                umems.push(boot::Mem::new(
                    GlobAddr::new_with(i as TileId, 0),
                    used,
                    true,
                ));

                // kernel memory
                let kmem =
                    MemMod::new(MemType::KERNEL, i as TileId, used, args::get().kmem as goff);
                used += args::get().kmem as goff;
                // configure EP to give us access to this range of physical memory
                ktcu::config_local_ep(1, |regs| {
                    ktcu::config_mem(
                        regs,
                        KERNEL_ID,
                        kmem.addr().tile(),
                        kmem.addr().offset(),
                        kmem.capacity() as usize,
                        Perm::RW,
                    );
                });
                mem.add(kmem);

                // root memory
                mem.add(MemMod::new(
                    MemType::ROOT,
                    i as TileId,
                    used,
                    cfg::FIXED_ROOT_MEM as goff,
                ));
                used += cfg::FIXED_ROOT_MEM as goff;

                // user memory
                let user_size = core::cmp::min((1 << 30) - cfg::PAGE_SIZE as goff, avail);
                mem.add(MemMod::new(MemType::USER, i as TileId, used, user_size));
                umems.push(boot::Mem::new(
                    GlobAddr::new_with(i as TileId, used),
                    user_size - args::get().kmem as goff,
                    false,
                ));
            }
            else {
                let user_size =
                    core::cmp::min((1 << 30) - cfg::PAGE_SIZE as usize, tile.mem_size());
                mem.add(MemMod::new(
                    MemType::USER,
                    i as TileId,
                    0,
                    user_size as goff,
                ));
                umems.push(boot::Mem::new(
                    GlobAddr::new_with(i as TileId, 0),
                    user_size as goff,
                    false,
                ));
            }
            kmem_idx += 1;
        }
        else {
            if kmem_idx > 0 {
                panic!("All memory tiles have to be last");
            }

            LAST_TILE.set(i as TileId);

            if i > 0 {
                assert!(kernel_tile() == 0);
                utiles.push(boot::Tile::new(i as u32, *tile));
            }
        }
    }

    // write-back boot info
    let mut uoffset = addr.offset();
    uinfo.tile_count = utiles.len() as u64;
    uinfo.mem_count = umems.len() as u64;
    ktcu::write_slice(addr.tile(), uoffset, &[uinfo]);
    uoffset += size_of::<boot::Info>() as goff;
    uoffset += info.mod_count as goff * size_of::<boot::Mod>() as goff;

    // write-back user tiles
    ktcu::write_slice(addr.tile(), uoffset, &utiles);
    uoffset += uinfo.tile_count as goff * size_of::<boot::Tile>() as goff;

    // write-back user memory regions
    ktcu::write_slice(addr.tile(), uoffset, &umems);

    platform::KEnv::new(info, addr, mods, tiles)
}

pub fn init_serial(dest: Option<(TileId, EpId)>) {
    if envdata::get().platform == envdata::Platform::HW.val {
        let (tile, ep) = dest.unwrap_or((0, 0));
        let serial = GlobAddr::new(envdata::get().kenv + 16 * 1024 * 1024);
        let tile_modid = TCU::tileid_to_nocid(tile);
        ktcu::write_slice(serial.tile(), serial.offset(), &[
            tile_modid as u64,
            ep as u64,
        ]);
    }
    else if let Some(ser_tile) = user_tiles().find(|i| tile_desc(*i).isa() == TileISA::SERIAL_DEV)
    {
        if let Some((tile, ep)) = dest {
            ktcu::config_remote_ep(ser_tile, 4, |regs| {
                let act = kif::tilemux::ACT_ID as ActId;
                ktcu::config_send(regs, act, 0, tile, ep, cfg::SERIAL_BUF_ORD, UNLIM_CREDITS);
            })
            .unwrap();
        }
        else {
            ktcu::invalidate_ep_remote(ser_tile, 4, true).unwrap();
        }
    }
}

pub fn kernel_tile() -> TileId {
    envdata::get().tile_id as TileId
}
pub fn user_tiles() -> platform::TileIterator {
    platform::TileIterator::new(kernel_tile() + 1, LAST_TILE.get())
}

pub fn is_shared(tile: TileId) -> bool {
    platform::tile_desc(tile).is_programmable()
}

pub fn rbuf_tilemux(_tile: TileId) -> goff {
    cfg::TILEMUX_RBUF_SPACE as goff
}
